File: latest/week2/workshop4_m6/W04_M06_demo.R
# ============================================================
# Workshop 4 — Module 6: Correlation & Association Visualization
# R + Python demo — UNYT Tirana
# ============================================================
library(tidyverse); library(GGally); library(patchwork); dir.create("output",showWarnings=FALSE)
apps <- read_csv("googleplaystore.csv") |>
  mutate(Reviews=as.numeric(Reviews), Installs=str_remove_all(Installs,"[+,]") |> as.numeric(),
         Price=str_remove(Price,"\\$") |> as.numeric()) |>
  filter(!is.na(Rating), Reviews>0) |>
  mutate(log_reviews=log10(Reviews+1), log_installs=log10(Installs+1))

# 1. ANSCOMBE'S QUARTET
p_ans <- datasets::anscombe |> pivot_longer(everything(),
  names_to=c(".value","set"), names_pattern="(.)(.)" ) |>
  ggplot(aes(x=x, y=y)) + geom_point(color="#1565C0") +
  geom_smooth(method="lm", se=FALSE, color="#E53935", linewidth=0.8) +
  facet_wrap(~set, ncol=2) + theme_minimal(base_size=9) +
  labs(title="Anscombe's Quartet: r = 0.82 in all four")
ggsave("output/anscombe.png", p_ans, width=7, height=6, dpi=300)

# 2. CORRELATION HEATMAP
num_cols <- apps |> select(Rating, log_reviews, log_installs, Price) |> drop_na()
corr <- cor(num_cols)
# Manual ggplot heatmap
corr_df <- as.data.frame(as.table(corr)) |> rename(Var1=Var1, Var2=Var2, r=Freq) |>
  mutate(lower=as.numeric(Var1)>=as.numeric(Var2))
p_heat <- ggplot(corr_df |> filter(lower), aes(x=Var1, y=Var2, fill=r)) +
  geom_tile(color="white", linewidth=1) +
  geom_text(aes(label=round(r,2)), size=4, fontface="bold") +
  scale_fill_gradient2(low="#E53935", mid="white", high="#1565C0", midpoint=0, limits=c(-1,1)) +
  theme_minimal() + labs(title="Correlation Heatmap (lower triangle)", fill="Pearson r") +
  theme(axis.text.x=element_text(angle=30, hjust=1))
ggsave("output/corr_heatmap.png", p_heat, width=6, height=5, dpi=300)

# 3. PAIRS PLOT
p_pairs <- ggpairs(apps |> select(Rating, log_reviews, Price, Type) |> drop_na() |> slice_sample(n=1000),
  mapping=aes(color=Type, alpha=0.3),
  lower=list(continuous=wrap("points",size=0.5)),
  diag=list(continuous=wrap("densityDiag")),
  upper=list(continuous=wrap("cor",size=3))) +
  scale_color_manual(values=c(Free="#1565C0",Paid="#E53935")) +
  theme_minimal(base_size=7)
ggsave("output/pairs_plot.png", p_pairs, width=8, height=8, dpi=300)

# 4. CATEGORICAL ASSOCIATION
cat_table <- apps |> filter(Category %in% c("GAME","FAMILY","TOOLS","BUSINESS","MEDICAL")) |>
  count(Category, Type) |> group_by(Category) |> mutate(pct=n/sum(n))
p_cat <- ggplot(cat_table, aes(x=Category, y=pct, fill=Type)) +
  geom_col(width=0.6, color="white") +
  scale_fill_manual(values=c(Free="#1565C0",Paid="#E53935")) +
  scale_y_continuous(labels=scales::percent) + coord_flip() +
  theme_minimal() + labs(title="100% Stacked: Type x Category")
ggsave("output/categorical_assoc.png", p_cat, width=7, height=4, dpi=300)

cat("\n── All W04-M06 R plots saved ──\n")

File: latest/week2/workshop4_m6/W04_M06_demo.py
"""W04-M06: Correlation & Association Visualization — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, seaborn as sns, os
from scipy.stats import pearsonr, spearmanr, kendalltau, chi2_contingency
os.makedirs("output",exist_ok=True); np.random.seed(42)

apps = pd.read_csv("googleplaystore.csv")
apps["Reviews"] = pd.to_numeric(apps["Reviews"], errors="coerce")
apps["Installs"] = apps["Installs"].str.replace(r"[+,]","",regex=True).pipe(pd.to_numeric,errors="coerce")
apps["Price"] = apps["Price"].str.replace("$","",regex=False).pipe(pd.to_numeric,errors="coerce")
apps = apps.dropna(subset=["Rating","Reviews"]).query("Reviews > 0")
apps["log_reviews"] = np.log10(apps["Reviews"]+1)
apps["log_installs"] = np.log10(apps["Installs"]+1)

# 1. CORRELATION HEATMAP
num_cols = ["Rating","log_reviews","log_installs","Price"]
corr = apps[num_cols].corr()
mask = np.triu(np.ones_like(corr, dtype=bool))
fig,ax = plt.subplots(figsize=(7,6))
sns.heatmap(corr, mask=mask, annot=True, fmt=".2f", cmap="RdBu_r", vmin=-1, vmax=1, center=0,
    square=True, linewidths=0.5, cbar_kws={"shrink":0.8,"label":"Pearson r"}, ax=ax)
ax.set_title("Correlation Heatmap (lower triangle)", fontweight="bold")
plt.tight_layout(); plt.savefig("output/corr_heatmap_py.png", dpi=300); plt.close()

# 2. PAIRS PLOT
sub = apps[["Rating","log_reviews","Price","Type"]].dropna().sample(800,random_state=42)
g = sns.pairplot(sub, hue="Type", palette={"Free":"#1565C0","Paid":"#E53935"},
    diag_kind="kde", plot_kws={"s":10,"alpha":0.3}, height=2)
g.fig.suptitle("Pairs Plot: scatter + KDE + hue", fontweight="bold", y=1.02)
plt.savefig("output/pairs_py.png", dpi=300, bbox_inches="tight"); plt.close()

# 3. THREE COEFFICIENTS
r_p, _ = pearsonr(apps["log_reviews"], apps["Rating"])
r_s, _ = spearmanr(apps["log_reviews"], apps["Rating"])
r_k, _ = kendalltau(apps["log_reviews"], apps["Rating"])
print(f"log(Reviews) vs Rating: Pearson={r_p:.3f}, Spearman={r_s:.3f}, Kendall={r_k:.3f}")

# 4. CATEGORICAL: chi-squared + Cramer's V
cats5 = apps.query("Category in ['GAME','FAMILY','TOOLS','BUSINESS','MEDICAL']")
ct = pd.crosstab(cats5["Category"], cats5["Type"])
chi2, p, dof, exp = chi2_contingency(ct)
n_obs = ct.sum().sum(); k = min(ct.shape)-1
cramers_v = np.sqrt(chi2 / (n_obs * k))
print(f"Chi2={chi2:.1f}, p={p:.4f}, Cramer's V={cramers_v:.3f}")

# 5. CATEGORICAL VIZ
ct_pct = ct.div(ct.sum(axis=1), axis=0)
fig,ax = plt.subplots(figsize=(7,4))
ct_pct.plot.barh(stacked=True, ax=ax, color={"Free":"#1565C0","Paid":"#E53935"}, edgecolor="white")
ax.set_xlabel("%"); ax.set_title(f"100% Stacked: Type x Category (Cramer's V = {cramers_v:.2f})", fontweight="bold")
ax.legend(fontsize=7); ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/categorical_py.png", dpi=300); plt.close()

print("All W04-M06 Python plots saved")
